### ISSUE TITLE
Documentation for `fromEnvironment` constructors is very vague about what environment they refer to

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#53147]
It is easy to believe, from the way the documentation is written:

* https://master-api.flutter.dev/flutter/dart-core/String/String.fromEnvironment.html
* https://master-api.flutter.dev/flutter/dart-io/Platform/environment.html

...that `String.fromEnvironment` refers to the same environment as `Platform.environment`, or, at least, refers to the shell environment of the compiler, or some such.

As far as I can tell, however, it only refers to `-D` defines given to the compiler.

We should clarify all the documentation that refers to "environment" to explain exactly what it means by the term, how to see what's defined there, how to add things there, how this differs by platform, and to have cross-references from one to the other to explain how to do things when you want the other one.

**Comment by mnordine** [comment]
Agreed, as I have been tripped up/confused by this before.

However, as this has nothing to do with Flutter, shouldn't the pages reference be:
- https://api.dart.dev/stable/3.0.7/dart-core/String/String.fromEnvironment.html
- https://api.dart.dev/stable/3.0.7/dart-io/Platform/environment.html

I assume that's where the Flutter pages get their content from? Or do they have completely separate sources (that would be unfortunate).

**Comment by Hixie** [comment]
Both are generated from the same underlying dart:core/dart:io library source code.

**Comment by parlough** [comment]
A bit of a related tangent, but "environment" is a loaded term and the name continues to cause a lot of confusion. Earlier this year we introduced a brief [Configuring apps with compilation environment declarations](https://dart.dev/guides/environment-declarations) article on dart.dev to resolve some of this confusion, and it was challenging how to go about clarifying this. The first paragraph and first note serve to do so, with "compilation environment declarations" being perhaps more accurate, but a mouthful. 

In https://github.com/dart-lang/sdk/issues/27585 @matanlurey proposed a new name for these functions, `fromDefine` which I really like and it matches Flutter's `--dart-define` and `--dart-define-from-file` options. Perhaps this confusion as well as those options and (from what I've seen, their relatively popular use) would be enough to justify that change now.

As for this particular issue, outside of a rename which would solve some of the confusion, I do think expanded docs and cross links would be great. Perhaps a link to the [dart.dev guide](https://dart.dev/guides/environment-declarations) I mentioned as well, which could likely use some of the same clarifications.

**Comment by matanlurey** [comment]
I would love if at minimum we created the new `.fromDefine` factories as an alias and deprecated the old ones (with a `dart fix`).

**Comment by Hixie** [comment]
i would be satisfied with the even simpler solution of just documenting what these APIs each mean by "environment" in their own API docs. :-)

**Comment by matanlurey** [comment]
It's a good step forward, sure, but it shouldn't prevent us from [doing the right thing](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#lazy-programming).

**Comment by parlough** [comment]
Yeah I agree that the doc improvements should be made either way. https://github.com/dart-lang/sdk/issues/27585 tracks the rename, but I hope this confusion can revitalize that discussion.

> (i would send a PR myself but I honestly am not sure what the actual answers are)

As for the actual answers, @lrhn provided great insight to the topic in the past.
